front/src/modules/timesStops/helpers/computeMargins.ts

```diff
@@ -1,41 +1,53 @@
 import type { TrainScheduleResult } from 'common/api/osrdEditoastApi';
-import type { TrainScheduleWithDetails } from 'modules/trainschedule/components/Timetable/types';
+import type {
+  TheoreticalMarginsRecord,
+  TrainScheduleWithDetails,
+} from 'modules/trainschedule/components/Timetable/types';
 import { ms2sec } from 'utils/timeManipulation';
 
 import { formatDigitsAndUnit } from './utils';
+import type { ScheduleEntry } from '../types';
 
-function getTheoreticalMargin(selectedTrainSchedule: TrainScheduleResult, pathStepId: string) {
-  if (selectedTrainSchedule.path.length === 0) {
+export function getTheoreticalMargins(selectedTrainSchedule: TrainScheduleResult) {
+  const selectedTrainScheduleMargins = selectedTrainSchedule.margins;
+  if (!selectedTrainScheduleMargins) {
     return undefined;
   }
-  // pathStep is starting point => we take the first margin
-  if (selectedTrainSchedule.path[0].id === pathStepId) {
-    return selectedTrainSchedule.margins?.values[0];
-  }
-  const theoreticalMarginBoundaryIndex = selectedTrainSchedule.margins?.boundaries?.findIndex(
-    (id) => id === pathStepId
-  );
-  if (
-    theoreticalMarginBoundaryIndex === undefined ||
-    theoreticalMarginBoundaryIndex < 0 ||
-    theoreticalMarginBoundaryIndex > selectedTrainSchedule.margins!.values.length - 2
-  ) {
-    return undefined;
-  }
-
-  return selectedTrainSchedule.margins!.values[theoreticalMarginBoundaryIndex + 1];
+  const theoreticalMargins: TheoreticalMarginsRecord = {};
+  let marginIndex = 0;
+  selectedTrainSchedule.path.forEach((step, index) => {
+    let isBoundary = index === 0;
+    if (step.id === selectedTrainSchedule.margins?.boundaries[marginIndex]) {
+      marginIndex += 1;
+      isBoundary = true;
+    }
+    theoreticalMargins[step.id] = {
+      theoreticalMargin: selectedTrainScheduleMargins.values[marginIndex],
+      isBoundary,
+    };
+  });
+  return theoreticalMargins;
 }
 
 function computeMargins(
+  theoreticalMargins: TheoreticalMarginsRecord | undefined,
   selectedTrainSchedule: TrainScheduleResult,
+  scheduleByAt: Record<string, ScheduleEntry>,
   pathStepIndex: number,
   pathItemTimes: NonNullable<TrainScheduleWithDetails['pathItemTimes']> // in ms
 ) {
   const { path, margins } = selectedTrainSchedule;
+  const pathStepId = path[pathStepIndex].id;
   if (
     !margins ||
-    (margins.values.length === 1 && margins.values[0] === '0%') ||
-    pathStepIndex === selectedTrainSchedule.path.length - 1
+    pathStepIndex === selectedTrainSchedule.path.length - 1 ||
+    !theoreticalMargins ||
+    !theoreticalMargins[pathStepId] ||
+    !(
+      scheduleByAt[pathStepId]?.arrival ||
+      scheduleByAt[pathStepId]?.stop_for ||
+      theoreticalMargins[pathStepId].isBoundary
+    )
```

Comment:
I think you can simplify like this:
```suggestion
  const schedule = scheduleByAt[pathStepId];
  const currentTheoreticalMargin = theoreticalMargins?.[pathStepId];
  if (
    !theoreticalMargins ||
    pathStepIndex === path.length - 1 ||
    !schedule ||
    !currentTheoreticalMargin?.isBoundary
  ) {
    return {
      theoreticalMargin: undefined,
      theoreticalMarginSeconds: undefined,
      calculatedMargin: undefined,
      diffMargins: undefined,
    };
  }
```

We don't display the margins if:
- the pathStep is the final pathStep
- the trainSchedule has no margins (and no theoreticalMargins)
- the pathStep has no schedule (if it has a schedule, it should have either an arrival or a stop_for)
- if a new margin is defined at this pathStep

Reply:
Great suggestion thanks!

(I renamed currentTheoreticalMargin into stepTheoreticalMarginInfo to avoid confusion with other variables that follow)